### ISSUE TITLE
Tips to save memory usage?

### ISSUE THREAD

**Issue by chconnor** [domlysz/BlenderGIS#101]
Hi! I am using BlenderGIS to import and view terrain from the USGS [nationalmap.gov](https://viewer.nationalmap.gov/basic) site. From their "3DEP" products, I convert .IMG files to .tif with gdal_translate, and then import via BlenderGIS as a "raster georeferenced world file"; I choose "As DEM" and subdivision "mesh".

This works great, but only for small files (3MB tif or so). For larger files (like most of those from the USGS which are 200MB, 300MB, etc.) memory is quickly exhausted and Blender will crash or hang my system, etc.

I have 16GB RAM and a video card... is there any hope for me to view these files?

I am a total beginner at this GIS stuff, so I might be doing something completely stupid.

Alternately, consider this a feature request to magically make BlenderGIS use less RAM, if it's possible. :-)

Thanks!

**Comment by domlysz** [comment]
are you able to open these kind of image directly in Blender's image editor (ie without the addon), if the system freeze it's a Blender issue. If not it's probably because your raster is coded with signed 16 bits depth and in this case the addon must convert it in float which can overflow the ram if the raster is too huge. Make sure to force the conversion to float32 with gdal_translate (add the option **-ot float32**)

I must admit I haven't designed the addon to handle very large dataset

**Comment by chconnor** [comment]
Thanks for the reply!

Yes, the .IMG files open quickly and easily in Blender's image editor. And sorry for my lack of clarity: yes, the hanging is due to my system's lack of swap space -- I just meant that it runs out of RAM (and in my case the system tar-pits so badly it must be power cycled). Certainly not the plugin's fault.

I did `gdal_translate -ot float32 in.img out.tif` as you suggested, but it did not change the result: runaway RAM usage.

I note that the results of `gmic -p` on the version done with `gdal_translate in.img out.tif` and the version done with `gdal_translate -ot float32 in.img out.tif` are identical -- it seems like gdal_translate is defaulting to float32, at least for these files.

Any easy/straightforward way to have the plugin chop the RAM-heavy task into smaller bites? It's probably harder than it sounds to implement, I know. :-)

Thanks for considering the issue!

**Comment by chconnor** [comment]
(and if a mesh of this size is just too big for my computer, I can accept that, too -- I don't know what to expect from these files...)

**Comment by domlysz** [comment]
ok, there are some situation where the addon must pre-process the file before open it in Blender and this step load all the data in the RAM. It happen when the raster is coded with signed 16bits integer and when the user check clip or fillnodata options, otherwise the tif file is just opened as it, so if blender can handle it, in theory it must work with the addon !

So maybe it comes from the subdivision process. I think the _mesh_ method is to blame here, it literally enter in edit mode and performs 2 loopcut to get one vertex per pixel. Try using the _subsurf_ method or set the option to none to see if the problem come from this.

hope it help, for may part I will try to test a large file today

**Comment by PayamT** [comment]
I agree with Dolmysz on using Subsurf method and you can always refine the surface quality by manually subdividing your data after import. As another strategy, preprocessing the data in some GIS software also always help. You can use GRASS or QGIS (if you want to avoid expensive Arcmap) to resample the raster to a slighty coarser resolution. This immensley reduces the file size and eases the import process. 


Sent from Yahoo Mail for iPhone


On Saturday, March 10, 2018, 3:57 AM, domlysz <notifications@github.com> wrote:


ok, there are some situation where the addon must pre-process the file before open it in Blender and this step load all the data in the RAM. It happen when the raster is coded with signed 16bits integer and when the user check clip or fillnodata options, otherwise the tif file is just opened as it, so if blender can handle it, in theory it must work with the addon !

So maybe it comes from the subdivision process. I think the mesh method is to blame here, it literally enter in edit mode and performs 2 loopcut to get one vertex per pixel. Try using the subsurf method or set the option to none to see if the problem come from this.

hope it help, for may part I will try to test a large file today

**Comment by chconnor** [comment]
@domlysz -- yes, the subsurf method works immediately, it's just much lower resolution. If you need a large file to test on, you could download the one I've been trying to use [here](http://caseyconnor.org/delme/test.tif).

@PayamT -- thanks -- is there a way to refine a subsurf import so that it is effectively full-resolution? Meaning, not just smoothing? I don't want to coarsen the data... I need all the resolution I can get from the file.

Thanks, all!

**Comment by PayamT** [comment]
Yes- after import, turn on edit mode, select all faces , apply subdivision (from toolbox) and turn back to object mode. Start from smaller subdivisions.


Sent from Yahoo Mail for iPhone


On Saturday, March 10, 2018, 1:39 PM, Casey Connor <notifications@github.com> wrote:


@domlysz -- yes, the subsurf method works immediately, it's just much lower resolution. If you need a large file to test on, you could download the one I've been trying to use here.

@PayamT -- thanks -- is there a way to refine a subsurf import so that it is effectively full-resolution? Meaning, not just smoothing? I don't want to coarsen the data... I need all the resolution I can get from the file.

Thanks, all!

**Comment by chconnor** [comment]
Great -- that works really well, thank you. Seems like I can get higher-res versions via that route even compared to imported as mesh (e.g. on smaller files that work both ways). Only issue is that the subdivision seems to leave some grid-like artifacts... meaning, with subdivisions high enough to remove the stair-stepping grid appearance, there is a larger-scale grid artifact that appears amongst the data... so the "mesh" import looks nicer overall. But it's a lot farther than I was getting before!

**Comment by PayamT** [comment]
Yes if you push the subdivision too far— more than the GIS resoulution — you will see those grids. There is a sweet spot in subdivision where you get the best surface quality while keeping the file size managable. 


Sent from Yahoo Mail for iPhone


On Saturday, March 10, 2018, 8:26 PM, Casey Connor <notifications@github.com> wrote:


Great -- that works really well, thank you. Seems like I can get higher-res versions via that route even compared to imported as mesh (e.g. on smaller files that work both ways). Only issue is that the subdivision seems to leave some grid-like artifacts... meaning, with subdivisions high enough to remove the stair-stepping grid appearance, there is a larger-scale grid artifact that appears amongst the data... so the "mesh" import looks nicer overall. But it's a lot farther than I was getting before!

**Comment by chconnor** [comment]
Ah, I'm getting it now. "subsurf" contains the underlaying image and represents it as deformation to the overlaying surface, which I can subdivide as desired. Hence the aliasing if I push the subdivision too far. "mesh" presumably generates a one-vertex-per-pixel mesh right off the bat. Thanks for the tutorial. :-)

It'd be great if the "mesh" import could magically use less RAM, as that seems like the simplest route to getting a 1-to-1 mesh, which is my goal. And since I apparently can hold the resulting mesh in memory when I go the subsurf route, maybe it's possible to make that happen.

Either way, thanks for the attention to it and the help!

**Comment by domlysz** [comment]
I think, commit 3ea1627bf36bef2d0343652c8bcd950751165bf9 now resolves this issue. It improve RAW DEM import mode by adding the ability to also build faces instead of producing only points cloud (this way is faster than performing the triangulation afterwards). So, RAW DEM import is now the normal way to get the expected 1-to-1 mesh. As DEM import (now renamed "As DEM texture") use in priority surbsurf+displacer combinaison, I've removed the loopcut hack which gave poor performance as pointed in this issue.

**Comment by chconnor** [comment]
Brilliant, thanks! It does seem to use much less memory now. I still can't open the ~250MB files at pixel step 1 (due to RAM usage), but I can open them at pixel step 2. They don't appear correctly once imported, but that may well be user error of some kind. It opens smaller files just fine. I'll continue poking at it.

Thanks for looking at it!

**Comment by domlysz** [comment]
ok, I'm interesting for a re-upload of this 250mb file. I think there are still some tricks to experiment for saving memory here.

**Comment by chconnor** [comment]
Sure! Here's a link: http://caseyconnor.org/pub/image/demo_gis.tif
~264MB. I have 12GB or so free RAM and open via the T-menu tab button "Import raster georeferenced..." button. As DEM texture, mesh, unchecked "apply on existing", unchecked "fill no data", step 1, WGS84 latlon. That will run out of RAM for me.

As mentioned, I can open it at pixel step 2 (16M faces). It gets close to running out of memory, but just before it does the memory usage drops back down. Takes a further 5 or 10m to open. The resulting mesh doesn't seem to be right, though. Again, could easily be user error on that, I don't have much experience in this area. The geographic scale is ~large in this file so that may be related. Haven't had a chance to explore this further yet. (It opens as a ~9" by ~9" by ~6 mile object. I scale the Z dimension to 1' and it appears as a flat square horizontal plane adjoining a similar flat vertical plane on one side, the top edge of which seems to be contoured from the data.)

(Side note: it would be handy for me if the import did not change the render mode to "texture" by default as it is much slower for seeing the results. I change it to "solid", no big deal, but just a note that the default is usually too slow on my machine. There could be excellent reasons to default to "texture" that I'm ignorant of, of course, just mentioning my experience. :-) )

Even if you work further magic to make the file open at step 1, I'm not sure my computer would handle it, as it can barely work with the step 2 file... ?

**Comment by chconnor** [comment]
Oops, just read your comment closer, sorry! (In my excitement I missed the distinction between raw dem and dem texture): I just imported as "Raw DEM" with "build faces" enabled and that worked at step 1! But the result also looked strange... the vertical scale was ~3000 feet, which is probably accurate to the data, but the x/y was still ~9" either direction.

I'll try with some other .tifs as well... whatever is going wrong is probably my fault...

**Comment by chconnor** [comment]
So that demo .tif that I linked is mostly null data with a strip of elevation at the bottom, though I'm not sure that explains what I saw. Anyway, I tried with a 402MB file (Raw DEM, build faces) and the import runs out of memory. You can download it here if it's helpful: http://caseyconnor.org/pub/image/demo_gis2.tif

**Comment by chconnor** [comment]
Thanks for the help!

I'm sure you are correct -- even opening at step 2 on these large meshes my Blender can barely handle it. I have no expectation that this script should be able to magically make Blender able to handle gigantic high-detail meshes. :-) I have been doing what you suggest: opening the .tif in something else and cropping to the area of interest, or opening at lower resolution, etc.

(The use of the large meshes can be handy if one is making a fly-through where it starts close up and pulls out to a wide shot, but I do understand your point: giant low-detail or smaller high-detail is going to work best in Blender.)

Thanks for the tip on QGIS! I had not found that. I've been using very cumbersome methods to investigate the data until now. Very useful!

demo2_gis.tif is hopefully a more useful example (less null data), in case it's helpful for improving memory usage.
